### PR TITLE
feat: openai whisper support

### PR DESCRIPTION
**Pull request by gaetansnl**
Still in WIP
- [x] linear
- [x] layernorm
- [x] attention
- [ ] ~constant matching (needed for constant in attention, and multiple model architecture support)~
- [x] benchmarks

### CODE REVIEW

**Comment by pommedeterresautee** [comment]
did a simple test, encoder optimization doesn't really matter:

```
encoder:  0.011400222778320312
decoder:  0.19341230392456055
```

**Comment by gaetansnl** [comment]
original attention code
time spend per shape type

**whisper-base**

Small shapes
15.1529
Large shapes
19.651

**whisper-medium**

Small shapes
56.60
Large shapes
78.6643

**Comment by gaetansnl** [comment]
`q.size(-2) == 1 and k.size(-2) > 50` for attention_reference, else attention_forward

**with replacement of attention**

```
test/test_torchdynamo.py ..
test/test_torchdynamo.py::test_whisper_hf
Name                        Median (CUDA)    Mean (CUDA)      Min (CUDA)       Max (CUDA)       Median           Mean             Min              Max
--------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------
test_whisper_hf[optimized]  248.3679 (2.23)  248.3679 (2.23)  248.3679 (2.23)  248.3679 (2.23)  253.9064 (2.16)  253.9064 (2.16)  253.9064 (2.16)  253.9064 (2.16)
test_whisper_hf[reference]  553.4851 (1.0)   553.4851 (1.0)   553.4851 (1.0)   553.4851 (1.0)   548.7756 (1.0)   548.7756 (1.0)   548.7756 (1.0)   548.7756 (1.0)
```
**without (still linear replacement, but no layernorm replacement)**
```
test/test_torchdynamo.py ..
test/test_torchdynamo.py::test_whisper_hf
Name                        Median (CUDA)    Mean (CUDA)      Min (CUDA)       Max (CUDA)       Median           Mean             Min              Max
--------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------
test_whisper_hf[optimized]  277.5296 (1.98)  277.5296 (1.98)  277.5296 (1.98)  277.5296 (1.98)  285.3432 (1.95)  285.3432 (1.95)  285.3432 (1.95)  285.3432 (1.95)
test_whisper_hf[reference]  549.0862 (1.0)   549.0862 (1.0)   549.0862 (1.0)   549.0862 (1.0)   556.6385 (1.0)   556.6385 (1.0)   556.6385 (1.0)   556.6385 (1.0)

```

**Comment by gaetansnl** [comment]
openai/whisper-medium
min_length=25, max_length=25, num_beams=2

**Without condition** so always attention_forward, everything enabled,  but no layernorm replacement
```
test/test_torchdynamo.py ..
test/test_torchdynamo.py::test_whisper_hf
Name                        Median (CUDA)    Mean (CUDA)      Min (CUDA)       Max (CUDA)       Median           Mean             Min              Max
--------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------
test_whisper_hf[optimized]  266.6812 (2.06)  266.6812 (2.06)  266.6812 (2.06)  266.6812 (2.06)  271.8331 (2.04)  271.8331 (2.04)  271.8331 (2.04)  271.8331 (2.04)
test_whisper_hf[reference]  549.4655 (1.0)   549.4655 (1.0)   549.4655 (1.0)   549.4655 (1.0)   554.8252 (1.0)   554.8252 (1.0)   554.8252 (1.0)   554.8252 (1.0)
```

Slow as expected

**Comment by gaetansnl** [comment]
split implementation at the time of writing

```
test/test_torchdynamo.py ..                                                                                                                                                                                                       [100%]
test/test_torchdynamo.py::test_whisper_hf
Name                        Median (CUDA)    Mean (CUDA)      Min (CUDA)       Max (CUDA)       Median           Mean             Min              Max
--------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------
test_whisper_hf[optimized]  244.4717 (2.23)  244.4717 (2.23)  244.4717 (2.23)  244.4717 (2.23)  251.3681 (2.17)  251.3681 (2.17)  251.3681 (2.17)  251.3681 (2.17)
test_whisper_hf[reference]  544.7885 (1.0)   544.7885 (1.0)   544.7885 (1.0)   544.7885 (1.0)   546.2297 (1.0)   546.2297 (1.0)   546.2297 (1.0)   546.2297 (1.0)
```

**Comment by gaetansnl** [comment]
image
image

File: src/kernl/implementations/attention.py [hunk 1]
```diff
@@ -57,7 +57,7 @@ def attention_reference(
         M = torch.tril(torch.ones((size_m, size_n), device="cuda"))
         p = torch.where(M == 0, float("-inf"), p)
     p = torch.nn.functional.softmax(p, dim=-1)
-    ref_out = torch.matmul(p, v, out=output)
+    ref_out = torch.matmul(p.to(v.dtype), v, out=output)
```
Comment (gaetansnl) [inline src/kernl/implementations/attention.py:60]: I added this because AAM not working inside wrapper, what do you think @pommedeterresautee

Comment (pommedeterresautee) [reply src/kernl/implementations/attention.py:60]: ok to me

**Review by pommedeterresautee — APPROVED** [review]
lgtm
test pass
```
pytest test/test_torchdynamo.py -k "optimized" 

====================================================================================== 107 passed, 161 deselected in 2655.59s (0:44:15) ======================================================================================
```

Same for whisper test

**Pull request merged**